sdk/containerregistry/azure-containerregistry/azure/containerregistry/_exchange_client.py

```diff
@@ -0,0 +1,95 @@
+# coding=utf-8
+# ------------------------------------
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+# ------------------------------------
+import re
+from typing import TYPE_CHECKING
+
+from azure.core.pipeline.policies import SansIOHTTPPolicy
+
+from ._generated import ContainerRegistry
+from ._helpers import _parse_challenge
+from ._user_agent import USER_AGENT
+
+if TYPE_CHECKING:
+    from azure.core.credentials import TokenCredential
+    from azure.core.pipeline import PipelineRequest, PipelineResponse
+    from typing import Dict, Any, List
+
+
+class ExchangeClientAuthenticationPolicy(SansIOHTTPPolicy):
+    """Authentication policy for exchange client that does not modify the request"""
+
+    def on_request(self, request):
+        # type: (PipelineRequest) -> None
+        pass
+
+    def on_response(self, request, response):
+        # type: (PipelineRequest, PipelineResponse) -> None
+        pass
+
+
+class ACRExchangeClient(object):
+    """Class for handling oauth authentication requests
+
+    :param endpoint: Azure Container Registry endpoint
+    :type endpoint: str
+    :param credential: Credential which provides tokens to authenticate requests
+    :type credential: :class:`azure.core.credentials.TokenCredential`
+    """
+
+    BEARER = "Bearer"
+    AUTHENTICATION_CHALLENGE_PARAMS_PATTERN = re.compile('(?:(\\w+)="([^""]*)")+')
+
+    def __init__(self, endpoint, credential, **kwargs):
+        # type: (str, TokenCredential, Dict[str, Any]) -> None
+        if not endpoint.startswith("https://") and not endpoint.startswith("http://"):
+            endpoint = "https://" + endpoint
+        self._endpoint = endpoint
+        self._credential_scopes = "https://management.core.windows.net/.default"
+        self._client = ContainerRegistry(
+            credential=credential,
+            url=endpoint,
+            sdk_moniker=USER_AGENT,
+            authentication_policy=SansIOHTTPPolicy(),
+            credential_scopes=kwargs.pop("credential_scopes", self._credential_scopes),
+            **kwargs
+        )
+        self._credential = credential
+
+    def get_acr_access_token(self, challenge):
```

Comment:
These are not public APIs right?
But they're still using a pipeline - so should still probably still support kwargs and have these passed in?

Reply:
I added them in, but how can a kwarg be passed in to these calls if they're not called directly by the user? Should the **kwarg be passed in from the `authentication_policy` that calls them?

Reply:
Yeah, from the request context.
Though you could open an issue to address that in a subsequent preview.

Reply:
Will do, I'll add a tracking issue to add tests to make sure the kwargs are being passed through. #17735